### DISCUSSION TITLE
InteractiveAutoRenderMode not working  with  Oqtane 5.1

### DISCUSSION THREAD

**Discussion by qunshen** [oqtane/oqtane.framework#4059] [General]
I played with Oqtane5.1 , and found that when you set InteractiveAutoRenderMode  render model in site setting, the wholw site runs the same way as Webassembly mode. the UI displays only after runtime downloaded.  with InteractiveAutoRenderMode  ,the UI are supposed tp display quickly when connects to server before the runtime downloaded!

**Comment by sbwalker** [comment]
@qunshen I believe this is because Oqtane is setting StreamRendering in the Routes and Head component - which is necessary to get Static Blazor to render content (if the StreamRendering is removed the framework does not render any content). This issue has been logged with Microsoft - https://github.com/dotnet/aspnetcore/issues/54157 and I am hopeful will be resolved eventually. In the meantime your observation is correct that Auto render mode does not work as intended - but we are still going to release 5.1 this week as all of the other render modes are functioning properly.

**Comment by qunshen** [comment]
from my perpective, auto render mode is the most userful  feature in real world application。 I also look into the default project template of  Balzor WebApp and the auto render mode is functional as intended! Thanks!

**Reply by sbwalker** [reply]
You can correct that Auto render mode works fine in a simplistic "Hello World" application which does not use any web services, database, etc...

**Reply by sbwalker** [reply]
As a mentioned above, I am hopeful that Microsoft will be able to provide some assistance to explain why real world applications refuse to render content in Blazor in .NET 8 under a variety of scenarios (and StreamRendering seems to "resolve" the problem)

**Comment by qunshen** [comment]
It appears that auto render mode has something to do the routing! if you use "@page "/xxx"", it works as intended!

**Reply by sbwalker** [reply]
Oqtane does not support rouitable components (ie. components using ```@page```). This is because routes are dynamic in Oqtane - ie. they are maintained in a database.

**Comment by qunshen** [comment]
I have an ideal: can we simulate the "@page /xxx" by fall back all route to one page which has only one component--siterouter?  this way,oqtance only has one page using "@page/"

**Reply by sbwalker** [reply]
Oqtane is already using a fall back route for its custom router - see Oqtane.Server\Extensions\ComponentEndpointRouteBuilderExtensions.cs

**Comment by qunshen** [comment]
this is at Server. if we can find a way to fallback to one page at client, we might able to using "@page" to specify router while to use Oqtane's dynamical nature.

**Reply by sbwalker** [reply]
The SiteRouter is used in ALL routing scenarios in Oqtane ie. all render modes and hosting models.

**Reply by sbwalker** [reply]
@qunshen if you want to experiment with a more a more simplistic example, I would encourage you to look at the Oqtane SSR POC - https://github.com/oqtane/OqtaneSSR . It was created to validate that Oqtane could support Static Server-Side Rendering but it also uses the same routing concepts as the Oqtane framework (ie. fallback/SiteRouter).

**Comment by qunshen** [comment]
Yes. but Oqtane doest not use ."@page".  what I an suggeting is that if we can use "@page" to define one page and all dynamical generated pages fallback to the page at client side. this way ,Auto render mode might work.

**Reply by sbwalker** [reply]
@qunshen as I mentioned above, please feel free to experiment using the Oqtane SSR POC code base to try out your ideas. 5.1.0 is being released in the next few days so any enhancements would need to go into 5.1.1.

**Comment by qunshen** [comment]
ok! my point is that making oqtane a more "@page/"  type application, while perserve its dynamcal nature! I will do some experiment!

**Reply by sbwalker** [reply]
supporting ```@page``` routes is a theoretical solution... but the actual problem is how to make Auto render mode behave as expected in Oqtane. It is possible that ```@page``` routes may solve the problem, but it is also possible that the solution might be related to removing StreamRendering, or something else entirely. I am just saying this to ensure we focus on the actual problem.

**Comment by sbwalker** [comment]
Link to the Blazor default router which includes a comment:

https://github.com/dotnet/aspnetcore/blob/5d31d5ed82270db90c375b78ba4f99025f0aef56/src/Components/Components/src/Routing/Router.cs#L200

            // Other routers shouldn't provide RouteData, this is specific to our router component
            // and must abide by our syntax and behaviors.
            // Other routers must create their own abstractions to flow data from their SSR routing
            // scheme to their interactive router

**Comment by thabaum** [comment]
Hello everyone.  Here is some more interesting behavior I have observed while in Interactive Client/Auto modes and not found in any static mode from the HTML Module itself and changing pages.

While in Interactive Mode with Client or Auto modes selected when you change from the homepage which has the shared secured content component also to the Private page which has this same component, the content loaded you can see get refreshed which maybe a clue.  I believe changing render modes from server to web assembly and back again is not supported, once an interactive type has been selected I believe everything after that part of the component should inherit it.  This behavior is not experience in server or any static modes. 

For example if you set the HTML module component override to Interactive.Server and the app is running its component wrapper as Interactive.Client or Interactive.Auto there might be a boundary issue here we might be crossing causing an issue as well potentially?

**Comment by sbwalker** [comment]
@qunshen I noticed in the asp.net repo on GitHub that there were a number of major bugs with Auto render mode in the initial .NET 8 release. You should install .NET SDK 8.0.4 in your environment to see if it resolves your earlier Auto render mode issue.

**Reply by qunshen** [reply]
ok ! thank

**Comment by qunshen** [comment]
it appears works!
